include/unifex/when_all.hpp

```diff
@@ -213,19 +213,39 @@ struct _op<Receiver, Senders...>::type {
 
   void start() noexcept {
     stopCallback_.construct(
-        get_stop_token(receiver_), cancel_operation{stopSource_});
+        get_stop_token(receiver_),
+        cancel_operation<Receiver, Senders...>{*this});
     ops_.start();
   }
 
-  private:
+  void request_stop() noexcept {
+    // mark callback as running (own deliver_result)
+    auto oldCount = refCount_.fetch_add(1, std::memory_order_release);
+    UNIFEX_ASSERT(!(oldCount & callback_running_bit));
+    stopSource_.request_stop();
+    // unmark (deliver_result back to operation)
+    oldCount = refCount_.fetch_sub(1, std::memory_order_acq_rel);
```

Comment:
If I've understood @MarkCMann's comments below, I think we can relax these atomic operations, too.
```suggestion
    auto oldCount = refCount_.fetch_add(1, std::memory_order_relaxed);
    UNIFEX_ASSERT(!(oldCount & callback_running_bit));
    stopSource_.request_stop();
    // unmark (deliver_result back to operation)
    oldCount = refCount_.fetch_sub(1, std::memory_order_relaxed);
```

The buggy scenario this diff fixes is that a single thread is running `request_stop()`, which synchronously invokes a stop callback that synchronously completes the last of the `when_all`'s child operations.  In that scenario, the old code would destroy the `when_all`'s operation state out from under the `stopSource_` while it's still processing stop callbacks.  In this scenario, the new code defers operation state destruction until we invoke `set_done` on line 231.

In the buggy scenario in the new code, we have the following sequence, all on the same thread:
 1. _Receiver_ requests stop, invoking `request_stop()`,
 2. `request_stop()` makes `refCount_` odd and invokes `stopSource_.request_stop()`,
 3. `stopSource_.request_stop()` invokes stop callbacks,
 4. some stop callback invokes the last `element_complete()`,
 5. `element_complete()` decrements `refCount_` from 3 to 1 and invokes `deliver_result()`,
 6. `deliver_result()` destroys the stop callback (doing not much because it's still running on the current thread),
 7. `deliver_result()` observes `refCount_` is 1 and returns, unwinding back up to `request_stop()`.
 8. `request_stop()` decrements `refCount_` from 1 to 0 and invokes `set_done`.

In this flow, everything happens on the same thread so no synchronization is necessary between `request_stop()` and `element_complete()` or `deliver_result()`.  We still do require that the thread observe the results of all previously-completed operations, but that's accomplished by the load-acquire in `deliver_result()`, which happens-before the relaxed subtraction in `request_stop()`.

There are a few other scenarios:
 1. _Receiver_ never requests stop;
 2. _Receiver_ requests stop on thread T1 and this completes before the `when_all`'s final child operation completes on thread T2;
 3. _Receiver_ requests stop on T1 while `when_all`'s final child is busy completing on T2.

Scenario 1 is trivial—there's no race.

In scenario 2, the stop callback increments and decrements `refCount_` before the `when_all`'s final child drops `refCount_` to zero.  There are no memory effects on the stop-requesting thread that need to be observed by the final child so it's enough to ensure that refcount manipulations are well-ordered, which is provided by relaxed ordering.

In scenario 3, the stop callback is racing with `when_all` completion and the race is resolved by destroying the stop callback in `deliver_result()`.  Either the stop callback is destroyed:
 - _before_ it is invoked and will thus never be invoked;
 - _after_ it was invoked and we're back in scenario 2; or
 - _while_ it's being invoked, in which case destruction blocks, forcing us into scenario 2 with a full memory barrier.